cloudsql/postgres/database-sql/connect_connector.go

```diff
@@ -0,0 +1,93 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// [START cloud_sql_postgres_databasesql_connect_connector]
+package cloudsql
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"log"
+	"net"
+	"os"
+
+	"cloud.google.com/go/cloudsqlconn"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/stdlib"
+)
+
+func connectWithConnector() (*sql.DB, error) {
+	mustGetenv := func(k string) string {
+		v := os.Getenv(k)
+		if v == "" {
+			log.Fatalf("Warning: %s environment variable not set.\n", k)
+		}
+		return v
+	}
+
+	var (
+		// Either a DB_USER or a DB_IAM_USER should be defined. If both are
+		// defined, DB_IAM_USER takes precedence.
+		dbUser                 = os.Getenv("DB_USER")                   // e.g. 'my-db-user'
+		dbIAMUser              = os.Getenv("DB_IAM_USER")               // e.g. 'sa-name@project-id.iam'
+		dbPwd                  = mustGetenv("DB_PASS")                  // e.g. 'my-db-password'
+		dbName                 = mustGetenv("DB_NAME")                  // e.g. 'my-database'
+		instanceConnectionName = mustGetenv("INSTANCE_CONNECTION_NAME") // e.g. 'project:region:instance'
+		usePrivate             = os.Getenv("PRIVATE_IP")
+	)
+	if dbUser == "" && dbIAMUser == "" {
+		log.Fatal("Wawrning: One of DB_USER or DB_IAM_USER must be defined")
+	}
+
+	dsn := fmt.Sprintf("user=%s password=%s database=%s", dbUser, dbPwd, dbName)
+	config, err := pgx.ParseConfig(dsn)
+	if err != nil {
+		return nil, err
+	}
+	config.DialFunc = func(ctx context.Context, network, instance string) (net.Conn, error) {
+		if dbIAMUser != "" {
+			// [START cloud_sql_postgres_databasesql_auto_iam_authn]
+			d, err := cloudsqlconn.NewDialer(ctx, cloudsqlconn.WithIAMAuthN())
+			if err != nil {
+				return nil, err
+			}
+			return d.Dial(ctx, instanceConnectionName)
+			// [END cloud_sql_postgres_databasesql_auto_iam_authn]
+		}
+		if usePrivate != "" {
+			d, err := cloudsqlconn.NewDialer(ctx,
+				cloudsqlconn.WithDefaultDialOptions(cloudsqlconn.WithPrivateIP()))
```

Comment:
nit: prefer line breaking between `(` and `ctx`

(or initializing the dialer on a separate line)